File: fragments/solana_program_counter/solana_counter_interface.go
package solana_program_counter

import (
	"context"
	"encoding/binary"
	solana_program "ff/solana_program"
	"ff/solana_rpc"
	solana_transaction "ff/solana_transaction"
	"fmt"

	"github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"
)

func InitializeAccount(userKeypair solana.PrivateKey, programId solana.PublicKey) (solana.Signature, error) {
	client := solana_rpc.InitRpcClient()
	discriminator, err := solana_program.GetInstructionDiscriminator("initialize", "counter")
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to get instruction discriminator: %v", err)
	}
	counterPda := solana_program.GetProgramDerivedAddress(userKeypair.PublicKey(), programId, "counter")
	instr := solana.NewInstruction(
		programId,
		solana.AccountMetaSlice{
			&solana.AccountMeta{
				PublicKey:  userKeypair.PublicKey(),
				IsSigner:   true,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  counterPda,
				IsSigner:   false,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  solana.SystemProgramID,
				IsSigner:   false,
				IsWritable: false,
			},
		},
		discriminator,
	)
	tx, err := solana_transaction.CreateTxWithFeePayerAndLifetime(userKeypair, instr)
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to create transaction: %v", err)
	}

	sig, err := client.SendTransactionWithOpts(context.Background(), tx, rpc.TransactionOpts{PreflightCommitment: rpc.CommitmentConfirmed})
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to send transaction: %v", err)
	}

	return sig, nil
}

func GetCount(userKeypair solana.PrivateKey, programId solana.PublicKey) (uint64, error) {
	client := solana_rpc.InitRpcClient()
	counterPda := solana_program.GetProgramDerivedAddress(userKeypair.PublicKey(), programId, "counter")

	offset := uint64(8)
	length := uint64(8)
	opts := &rpc.GetAccountInfoOpts{
		Encoding:   solana.EncodingBase64,
		Commitment: rpc.CommitmentConfirmed,
		// offset removes the discriminator from the account data
		DataSlice: &rpc.DataSlice{Offset: &offset, Length: &length},
	}
	res, err := client.GetAccountInfoWithOpts(context.Background(), counterPda, opts)
	if err != nil {
		return 0, fmt.Errorf("failed to get account info: %w", err)
	}
	if res.Value == nil {
		return 0, fmt.Errorf("account not found")
	}

	data := res.Value.Data.GetBinary()
	if len(data) < 8 {
		return 0, fmt.Errorf("invalid data length: %d", len(data))
	}

	count := uint64(binary.LittleEndian.Uint64((data)))
	return count, nil
}

func IncrementCounter(userKeypair solana.PrivateKey, programId solana.PublicKey) (solana.Signature, error) {
	client := solana_rpc.InitRpcClient()
	discriminator, err := solana_program.GetInstructionDiscriminator("increment", "counter")
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to get instruction discriminator: %v", err)
	}
	counterPda := solana_program.GetProgramDerivedAddress(userKeypair.PublicKey(), programId, "counter")
	instr := solana.NewInstruction(
		programId,
		solana.AccountMetaSlice{
			&solana.AccountMeta{
				PublicKey:  counterPda,
				IsSigner:   false,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  userKeypair.PublicKey(),
				IsSigner:   true,
				IsWritable: true,
			},
		},
		discriminator,
	)
	tx, err := solana_transaction.CreateTxWithFeePayerAndLifetime(userKeypair, instr)
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to create transaction: %v", err)
	}

	sig, err := client.SendTransactionWithOpts(context.Background(), tx, rpc.TransactionOpts{PreflightCommitment: rpc.CommitmentConfirmed})
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to send transaction: %v", err)
	}

	return sig, nil
}

File: fragments/solana_program_counter/solana_counter_interface.py
from solders.pubkey import Pubkey
from solders.instruction import Instruction, AccountMeta
from solders.signature import Signature
from solders.keypair import Keypair
from solana.constants import SYSTEM_PROGRAM_ID
from construct import Struct, Int64ul
from fragments.solana_rpc import init_rpc_client
from fragments.solana_program import get_instruction_discriminator, get_program_derived_address
from fragments.solana_transaction import create_tx_with_fee_payer_and_lifetime


async def initialize_account(user_keypair: Keypair, program_id: Pubkey) -> Signature:
    discriminator = get_instruction_discriminator("initialize", "counter")
    counter_pda = get_program_derived_address(user_keypair.pubkey(), program_id, "counter")
    client = init_rpc_client()
    instruction = Instruction(
        program_id,
        discriminator,
        [
            AccountMeta(pubkey=user_keypair.pubkey(), is_signer=True, is_writable=True),
            AccountMeta(pubkey=counter_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
        ],
    )
    tx = await create_tx_with_fee_payer_and_lifetime(user_keypair, instruction)
    response = await client.send_transaction(tx)
    return response.value


async def get_count(user_keypair: Keypair, program_id: Pubkey) -> int:
    client = init_rpc_client()
    counter_pda = get_program_derived_address(user_keypair.pubkey(), program_id, "counter")
    response = await client.get_account_info(counter_pda)
    account_info = response.value

    if account_info is None:
        raise ValueError(f"Account {counter_pda} does not exist")

    # removes the discriminator from the account data
    raw_bytes = bytes(account_info.data)[8:]
    schema = Struct("count" / Int64ul)
    parsed = schema.parse(raw_bytes)

    return parsed.count


async def increment_counter(user_keypair: Keypair, program_id: Pubkey) -> Signature:
    discriminator = get_instruction_discriminator("increment", "counter")
    counter_pda = get_program_derived_address(user_keypair.pubkey(), program_id, "counter")
    client = init_rpc_client()
    instruction = Instruction(
        program_id,
        discriminator,
        [
            AccountMeta(pubkey=counter_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=user_keypair.pubkey(), is_signer=True, is_writable=True),
        ],
    )
    tx = await create_tx_with_fee_payer_and_lifetime(user_keypair, instruction)
    response = await client.send_transaction(tx)
    return response.value
